yjit/src/core.rs

```diff
@@ -4158,7 +4158,20 @@ pub fn invalidate_block_version(blockref: &BlockRef) {
     }
 
     // For each incoming branch
-    for branchref in block.incoming.0.take().iter() {
+    let mut incoming_branches = block.incoming.0.take();
+
+    // An adjacent branch will write into the start of the block being invalidated, possibly
+    // overwriting the block's exit. If we run out of memory after doing this, any subsequent
+    // incoming branches we rewrite won't be able use the block's exit as a fallback when they
+    // are unable to generate a stub. To avoid this, if there's an incoming branch that's
+    // adjacent to the invalidated block, make sure we process it last.
+    incoming_branches.sort_by_key(|branchref| {
+        let branch = unsafe { branchref.as_ref() };
+        let target_next = block.start_addr == branch.end_addr.get();
+        target_next
+    });
```

Comment:
This feels wrong.

Rather than sorting, I was originally hoping to do two passes over the list of branches, ignoring the adjacent branch on the first pass.

But I also didn't want to duplicate the code inside the loop, and everything I tried for reuse (inner fn, closure) felt uglier than sorting.

Reply:
Now that I've written that, finding the index of the adjacent branch (if any) and swapping it to the end of the slice would be more efficient than sorting. Let me try that...